File: api/internal/model/reading.go
package model

import "time"

// Reading is the shape of one soil test.
type Reading struct {
	ReadingID   int       `json:"reading_id"`
	RawLat      float64   `json:"raw_lat"`
	RawLng      float64   `json:"raw_lng"`
	AnchorID    *int      `json:"anchor_id"`    // *int = can be empty (null)
	DistrictID  *int      `json:"district_id"`  // *int = can be empty (null)
	N           float64   `json:"n"`
	P           float64   `json:"p"`
	K           float64   `json:"k"`
	PH          float64   `json:"ph"`
	Crop        string    `json:"crop"`
	Stage       string    `json:"stage"`
	TargetYield float64   `json:"target_yield"`
	Area        float64   `json:"area"`
	CreatedAt   time.Time `json:"created_at"`
	QualityFlag string    `json:"quality_flag"`
}

File: mobile/src/types/reading.ts
// What we SEND to the server when creating a reading.
// Note: no id, no anchor, no district. The server makes those.
export interface NewReading {
  raw_lat: number;
  raw_lng: number;
  n: number;
  p: number;
  k: number;
  ph: number;
  crop: string;
  stage: string;
  target_yield: number;
  area: number;
}

// What the server SENDS BACK.
// "extends NewReading" means: take all the fields above, and add these.
// This saves us from typing the same 10 fields twice.
export interface Reading extends NewReading {
  reading_id: number;
  anchor_id: number | null;    // the | means "or". This can be empty.
  district_id: number | null;  // empty if the point is outside all districts
  created_at: string;
  quality_flag: string;
}

// The small answer from POST /readings.
export interface CreateReadingResponse {
  reading_id: number;
}
